backend/utils/sse.ts

```diff
@@ -0,0 +1,40 @@
+import { Response } from 'express';
+import logger from '../logger';
+
+type Client = {
+  id: string;
+  res: Response;
+};
+
+class SSEManager {
+  private clients: Map<string, Client> = new Map();
+
+  addClient(clientId: string, res: Response) {
+    res.setHeader('Content-Type', 'text/event-stream');
+    res.setHeader('Cache-Control', 'no-cache');
+    res.setHeader('Connection', 'keep-alive');
+    res.flushHeaders(); // flush the headers to establish the SSE connection
+
+    this.clients.set(clientId, { id: clientId, res });
+    logger.warn({ clientId }, 'Client connected');
+
+    res.on('close', () => {
+      this.removeClient(clientId);
+    });
+  }
+
+  removeClient(clientId: string) {
+    this.clients.delete(clientId);
+  }
+
+  broadcast(event: string, data: any) {
+    const payload = `event: ${event}\ndata: ${JSON.stringify(data)}\n\n`;
+    logger.warn({ payload }, 'Broadcasting to SSE clients');
+
+    this.clients.forEach((client) => {
+      client.res.write(payload);
```

Comment:
Might be nice to log if this errors. I think that can be handled via a callback here (`client.res.write(payload, (err) => { /* do something */))`, or possibly an event like `res.on('error', () => {})`

Note, I tried the first method myself and it might need additional logic, because it triggered even on a successful `200` response.